Add Anaplan Sink plugin as a SparkSink plugin.

niceyang opened the pull request.

niceyang committed:
Add Anaplan Sink plugin as a SparkSink plugin.

albertshau commented on src/main/java/com/google/planningworks/cdap/plugins/sink/AnaplanSink.java:
```diff
@@ -0,0 +1,157 @@
+          new URI(config.getAuthServiceLocation()));
+        AnaplanService.setTaskMetadata(anaplanConfig);
+
+        OutputStream uploadOutputStream = retrieveOutputStreamForServerFile(config);
```
Can you try pulling this function out into its own serializable class, and pass in the workspace, model, server file name, and chunk size into its constructor (instead of the PluginConfig). That way Spark doesn't need to serialize the PluginConfig, which will avoid triggering CDAP-17975 and avoid the secure macro error.

niceyang replied:
Thanks! Updated and tested. The secure macro error is no longer triggered.

albertshau commented on src/main/java/com/google/planningworks/cdap/plugins/sink/AnaplanSinkConfig.java:
```diff
@@ -0,0 +1,178 @@
+ * The config class for {@link AnaplanSink} that contains all properties that need to be filled in
+ * by the user when building a pipeline.
+ */
+class AnaplanSinkConfig extends PluginConfig {
```
looks like there's a lot of duplication between this and the other config, would be good to have some parent class that has the common properties/validation logic.

niceyang replied:
Yeah, that's a good idea. Updated.

albertshau commented on docs/AnaplanSink-sparksink.md:
```diff
@@ -0,0 +1,24 @@
+ -----------
+ Sink plugin for supporting data import from CDF to Anaplan server.
+
+ The placeholder with the same name as in the `Server file` filed must be predefined in the given Anaplan model.
```
is 'filed' a typo? (filed -> file?)

albertshau replied:
It would also be good to document that this plugin does not write data in parallel so may take some time on larger datasets.

Are there any size restrictions for an Anaplan file?

niceyang replied:
Thanks, the doc is fixed. No there is no restrictions, but there is a general good practice that to limit the updated file not more than 1 GB.

albertshau commented on docs/AnaplanSink-sparksink.md:
```diff
@@ -0,0 +1,24 @@
+
+ The placeholder with the same name as in the `Server file` filed must be predefined in the given Anaplan model.
+
+ Properties
```
It would be good to have a section on supported data types. See https://github.com/data-integrations/google-cloud/blob/develop/docs/BigQueryTable-batchsink.md#data-type-mappings-from-cdap-to-bigquery for an example.

niceyang replied:
Thanks, updated!

albertshau commented on src/main/java/com/google/planningworks/cdap/plugins/sink/AnaplanSink.java:
```diff
@@ -0,0 +1,157 @@
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    FailureCollector collector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    config.validate(collector);
```
it would also be good to validate that all the fields in the input schema are supported types (assuming schema is not null). The input schema can be fetched from the StageConfigurer.

Does anaplan have any restrictions on column names? If so, would be good to validate the field names too.

niceyang replied:
Yeah, added. Thank you very much for the suggestion.

Checked with Anaplan team, no restriction on the column name as long as not character out of the ASCII range. Quite relaxing.

niceyang committed:
Address the comments from Albert.

This version does not pass the integration test, there is an issue in the schema validation.

niceyang committed:
Convert the nullableField to uonNullableField for the schema validation.

albertshau commented on docs/AnaplanSink-sparksink.md:
```diff
@@ -22,3 +24,24 @@
+| boolean        | bool          |
+| date           | date          |
+| decimal        | numeric       |
+| double / float | numeric       |
```
nit: put double, float, int, and long on separate rows.

Also, would be good to arrange the supported and unsupported ones alphabetically (boolean, date, decimal, ..., array, bytes, enum, ...)

niceyang replied:
Got, updated, thanks!

albertshau commented on src/main/java/com/google/planningworks/cdap/plugins/sink/SinkToAnaplanProcess.java:
```diff
@@ -0,0 +1,97 @@
+package com.google.planningworks.cdap.plugins.sink;
```
missing copyright

niceyang replied:
Oh, added. Thanks!

albertshau commented on src/main/java/com/google/planningworks/cdap/plugins/sink/AnaplanSinkConfig.java:
```diff
@@ -0,0 +1,121 @@
+    // validate logical types
+    if (logicalType != null) {
+      if (!SUPPORTED_LOGICAL_TYPES.contains(logicalType)) {
+        collector.addFailure(generateUnsupportedTypeErrorMessage(field, fieldSchema),
```
you can use 'withInputSchemaField()' to tell the UI to highlight the exact field that is causing problems. See https://github.com/data-integrations/google-cloud/blob/f0ae7de2be0d4798d771174c52aca19c4fb8ac0a/src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java#L262 for an example. 

This also lets you check more specific things in the unit test

niceyang replied:
Yeah, that's more clear, updated!

albertshau commented on src/test/java/com/google/planningworks/cdap/plugins/base/AnaplanPluginConfigTest.java:
```diff
@@ -0,0 +1,136 @@
+import static com.anaplan.client.AnaplanService.NAME_PASSWORD;
+import static com.anaplan.client.AnaplanService.NAME_SERVICE_LOCATION;
+import static com.anaplan.client.AnaplanService.NAME_USERNAME;
+import static com.anaplan.client.AnaplanService.NAME_WORKSPACE_ID;
```
style: didn't catch this before, but our projects don't import static variables. Instead, import AnaplanService and refer to them in this class as AnaplanService.NAME_MODEL_ID, etc

niceyang replied:
Got, updated!

albertshau commented on src/test/java/com/google/planningworks/cdap/plugins/sink/AnaplanSinkTest.java:
```diff
@@ -0,0 +1,126 @@
+  public void testUnsupportedSchema() {
+    schema = getUnsupportedSchema();
+    config.validate(failureCollector, schema);
+    Assert.assertEquals(/*expected =*/ 5, failureCollector.getValidationFailures().size());
```
better to build the set of fields that were incorrect from the ValidationFailures collected by the failureCollector and check that it's equal to the expected set of bad fields.

niceyang replied:
Thanks! Updated!

albertshau commented on docs/AnaplanSink-sparksink.md:
```diff
@@ -0,0 +1,47 @@
+| decimal        | numeric       |
+| double / float | numeric       |
+| int / long     | numeric       |
+| time           | unsupported   |
```
CDAP also has a timestamp and datetime type.

niceyang replied:
Oh, right, added. Thanks!

niceyang committed:
Address the comments from Albert.

albertshau reviewed (approved):
lgtm, please squash changes before merging.
